# nf-core/pangenome: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: https://10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: https://10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [GFAFFIX](https://github.com/marschall-lab/GFAffix)

  > Liao W, Asri M, Ebler J, Doerr D, Haukness M, Hickey G, Lu S, Lucas J K, Monlong J, Abel H J, Buonaiuto S, Chang X H, Cheng H, Chu J, Colonna V, Eizenga J M, Feng X, Fischer C, Fulton R S, Garg S, Groza C, Guarracino A, Harvey W T, Heumos S, Howe K, Jain M, Lu T, Markello C, Martin F J, Mitchell M W, Munson K M, Mwaniki M N, Novak A M, Olsen H E, Pesout T, Porubsky D, Prins P, Sibbesen J A, Tomlinson C, Villani F, Vollger M R, Human Pangenome Reference Consortium, Bourque G, Chaisson M J P, Flicek P, Phillippy A M, Zook J M, Eichler E E,Haussler D, Jarvis E D, Miga K H, Wang T, Garrison E, Marschall T, Hall I, Li H, Paten B. A Draft Human Pangenome Reference. Nature 617, 312–324 (2023). https://doi.org/10.1038/s41586-023-05896-x.

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

  > Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: https://10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

- [NET2COMMUNITIES](https://github.com/pangenome/pggb/blob/master/scripts/net2communities.py)

  > Traag, VA, Waltman, L & van Eck, NJ. From Louvain to Leiden: guaranteeing well-connected communities. Sci Rep 9, 5233 (2019). https://doi.org/10.1038/s41598-019-41695-z.

- [ODGI](https://pubmed.ncbi.nlm.nih.gov/35552372/)

  > Guarracino A, Heumos S, Nahnsen S, Prins P, Garrison E. ODGI: understanding pangenome graphs. Bioinformatics. Volume 38. Issue 13. July 2022. Pages 3319–3326. https://doi.org/10.1093/bioinformatics/btac308.

  > Heumos S, Guarracino A, Schmelzle J N M, Li J, Zhang Z, Nahnsen S, Prins P, Garrison E. Pangenome graph layout by Path-Guided Stochastic Gradient Descent. bioRxiv. https://www.biorxiv.org/content/10.1101/2023.09.22.558964v1.

- [PGGB](https://www.biorxiv.org/content/10.1101/2023.04.05.535718v1)

  > Garrison E, Guarracino A, Heumos S, Villani F, Bao Z, Tattini L, Hagmann J, Vorbrugg S, Marco-Sola S, Kubica S, Ashbrook D G, Thorell K, Rusholme-Pilcher R L, Liti G, Rudbeck E, Nahnsen S, Yang Z, Moses M N, Nobrega F L, Wu Y, Chen H, de Ligt J, Sudmant P H, Soranzo N, Colonna V, Williams R W, Prins P. Building pangenome graphs. bioRxiv. 2023.04.05.535718. doi: https://doi.org/10.1101/2023.04.05.535718.

  > Guarracino A, Buonaiuto S, de Lima L G, Potapova T, Rhie A, Koren S, Rubinstein B, Fischer C, Human Pangenome Reference Consortium, Gerton J L, Phillippy A M, Colonna V, Garrison E. Recombination between heterologous human acrocentric chromosomes. Nature 617, 335–343 (2023). https://doi.org/10.1038/s41586-023-05976-y.

  > Liao W, Asri M, Ebler J, Doerr D, Haukness M, Hickey G, Lu S, Lucas J K, Monlong J, Abel H J, Buonaiuto S, Chang X H, Cheng H, Chu J, Colonna V, Eizenga J M, Feng X, Fischer C, Fulton R S, Garg S, Groza C, Guarracino A, Harvey W T, Heumos S, Howe K, Jain M, Lu T, Markello C, Martin F J, Mitchell M W, Munson K M, Mwaniki M N, Novak A M, Olsen H E, Pesout T, Porubsky D, Prins P, Sibbesen J A, Tomlinson C, Villani F, Vollger M R, Human Pangenome Reference Consortium, Bourque G, Chaisson M J P, Flicek P, Phillippy A M, Zook J M, Eichler E E,Haussler D, Jarvis E D, Miga K H, Wang T, Garrison E, Marschall T, Hall I, Li H, Paten B. A Draft Human Pangenome Reference. Nature 617, 312–324 (2023). https://doi.org/10.1038/s41586-023-05896-x.

- [SAMTOOLS](https://pubmed.ncbi.nlm.nih.gov/19505943/)

  > Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R; 1000 Genome Project Data Processing Subgroup. The Sequence Alignment/Map format and SAMtools. Bioinformatics. 2009 Aug 15;25(16):2078-9. doi: https://10.1093/bioinformatics/btp352. Epub 2009 Jun 8. PMID: 19505943; PMCID: PMC2723002.

- [SEQWISH](https://pubmed.ncbi.nlm.nih.gov/36448683/)

  > Garrison E, Guarracino A. Unbiased pangenome graphs. Bioinformatics. 2023 Jan 1;39(1):btac743. doi: https://10.1093/bioinformatics/btac743. PMID: 36448683; PMCID: PMC9805579.

- [SMOOTHXG](https://www.biorxiv.org/content/10.1101/2023.04.05.535718v1)

  > Garrison E, Guarracino A, Heumos S, Villani F, Bao Z, Tattini L, Hagmann J, Vorbrugg S, Marco-Sola S, Kubica S, Ashbrook D G, Thorell K, Rusholme-Pilcher R L, Liti G, Rudbeck E, Nahnsen S, Yang Z, Moses M N, Nobrega F L, Wu Y, Chen H, de Ligt J, Sudmant P H, Soranzo N, Colonna V, Williams R W, Prins P. Building pangenome graphs. bioRxiv. 2023.04.05.535718. doi: https://doi.org/10.1101/2023.04.05.535718.

- [VCFLIB](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009123)

  > Garrison E, Kronenberg ZN, Dawson ET, Pedersen BS, Prins P (2022) A spectrum of free software tools for processing the VCF variant call format: vcflib, bio-vcf, cyvcf2, hts-nim and slivar. PLOS Computational Biology 18(5): e1009123. https://doi.org/10.1371/journal.pcbi.1009123.

- [VG](https://pubmed.ncbi.nlm.nih.gov/30125266/)

  > Garrison E et al. Variation graph toolkit improves read mapping by representing genetic variation in the reference. Nature biotechnology vol. 36,9 (2018): 875-879. doi: https://10.1038/nbt.4227.

- [WFMASH](https://github.com/waveygang/wfmash)

  > Guarracino A, Mwaniki N, Marco-Sola S, Garrison E. Wfmash: whole-chromosome pairwise alignment using the hierarchical wavefront algorithm. 2024. https://github.com/waveygang/wfmash.

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: https://10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: https://10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: https://10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: https://10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
